src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt

```diff
@@ -0,0 +1,90 @@
+package org.ptss.support.security
+
+
+import io.quarkus.security.UnauthorizedException
+import jakarta.enterprise.context.ApplicationScoped
+import jakarta.inject.Inject
+import jakarta.ws.rs.container.ContainerRequestContext
+import jakarta.ws.rs.container.ContainerRequestFilter
+import jakarta.ws.rs.container.ResourceInfo
+import jakarta.ws.rs.core.Context
+import jakarta.ws.rs.core.HttpHeaders
+import jakarta.ws.rs.core.NewCookie
+import jakarta.ws.rs.core.Response
+import jakarta.ws.rs.ext.Provider
+import org.eclipse.microprofile.config.inject.ConfigProperty
+
+
+@Provider
+@ApplicationScoped
+class AuthenticationFilter @Inject constructor(
+    @Context private val resourceInfo: ResourceInfo,
+    private val identityServiceClient: IdentityServiceClient,
+    private val jwtValidator: JwtValidator,
+    @ConfigProperty(name = "app.security.access-token-cookie-name", defaultValue = "access_token")
+    private val accessTokenCookieName: String,
+    @ConfigProperty(name = "app.security.refresh-token-cookie-name", defaultValue = "refresh_token")
+    private val refreshTokenCookieName: String
+) : ContainerRequestFilter {
+
+    override fun filter(requestContext: ContainerRequestContext) {
+        val annotation = getAuthenticationAnnotation(resourceInfo) ?: return
+        val accessToken = requestContext.cookies[accessTokenCookieName]?.value
+        val refreshToken = requestContext.cookies[refreshTokenCookieName]?.value
+
+        // Deny if refresh token is missing or invalid
+        if (refreshToken.isNullOrBlank() || !jwtValidator.isTokenValidAndNotBlank(refreshToken)) {
+            denyRequest(requestContext, annotation.message)
+        }
+
+        // If access token is valid, use it; otherwise, attempt to refresh
+        val tokenToUse = when {
+            jwtValidator.isTokenValidAndNotBlank(accessToken) -> accessToken!!
+            refreshToken.isNullOrBlank() -> {
+                denyRequest(requestContext, "Missing refresh token")
+                return
+            }
+            else -> tryRefreshAccessToken(requestContext, refreshToken)
+        }
+
+        // Check if the token has the required role
+        if (!jwtValidator.hasRequiredRole(tokenToUse, annotation.roles.toSet())) {
+            denyRequest(requestContext, annotation.message)
+        }
+    }
+
+    private fun tryRefreshAccessToken(
+        requestContext: ContainerRequestContext,
+        refreshToken: String
+    ): String = runCatching {
+        val newToken = identityServiceClient.refreshAccessToken(refreshToken)
+        val newAccessTokenCookie = createNewAccessTokenCookie(newToken)
+        requestContext.headers.add(HttpHeaders.SET_COOKIE, newAccessTokenCookie.toString())
+        newToken
+    }.getOrElse {
+        denyRequest(requestContext, "Unauthorized")
+        throw UnauthorizedException("Failed to refresh token")
```

Comment:
Is het nog nodig om een UnauthorizedException te gooien als je al `denyRequest(...)` hebt aangeroepen?
Ik vraag het omdat ik dit niet weet, ik was me ervan bewust dat `denyRequest` gelijk de code stopt en een Unauthorized message terugstuurt. In de vorige code doe je namelijk wel `denyRequest` en daarna ga je ervan uit dat de code daarna niet meer wordt uitgevoerd.